### PR TITLE
Login issues

### PR DESCRIPTION
**Pull request by orcema**
rebased on new master
origin Login/Account Link - issues #102

### CODE REVIEW

File: smart-home-provider/cloud/auth-provider.js [hunk 1]
```diff
@@ -201,7 +201,8 @@ Auth.registerAuth = function (app) {
     let user = SmartHomeModel.getUser(req.body.username, req.body.password);
     if (!user) {
       console.log('not a user', user);
-      return login(req, res);
+      return res.redirect(util.format('%s?client_id=%s&redirect_uri=%s&state=%s&response_type=code',
+      '/frontend', req.body.client_id, req.body.redirect_uri, req.body.state));
```
Comment (Fleker) [inline smart-home-provider/cloud/auth-provider.js]: `encodeURIComponent(req.body.redirect_uri)`

File: smart-home-provider/cloud/datastore.js [hunk 2]
```diff
@@ -285,6 +285,9 @@ Data.getProperties = function (uid, deviceIds = undefined) {
   let properties = {};
 
   if (!deviceIds) {
+    
+    if(!Data.hasOwnProperty(uid)) return properties;
```
Comment (Fleker) [inline smart-home-provider/cloud/datastore.js]: Can you put that into brackets and reformat the indentation?

File: smart-home-provider/cloud/smart-home-provider-cloud.js [hunk 3]
```diff
@@ -368,6 +368,12 @@ app.get('/smart-home-api/device-connection/:deviceId', function (request, respon
 // frontend UI
 app.set('jsonp callback name', 'cid');
 app.get('/getauthcode', function (req, resp) {
+  
+  /*forbid caching to force reload of getauthcode */
```
Comment (Fleker) [inline smart-home-provider/cloud/smart-home-provider-cloud.js]: Please add a space after the /*

File: smart-home-provider/cloud/smart-home-provider-cloud.js [hunk 4]
```diff
@@ -368,6 +368,12 @@ app.get('/smart-home-api/device-connection/:deviceId', function (request, respon
...
+  resp.set('Cache-Control','no-store, must-revalidate');
+  /*set correct mime type else browser will refuse to execute the script*/
```
Comment (Fleker) [inline smart-home-provider/cloud/smart-home-provider-cloud.js]: Please add a space after the /*

File: smart-home-provider/cloud/smart-home-provider-cloud.js [hunk 5]
```diff
@@ -368,6 +368,12 @@ app.get('/smart-home-api/device-connection/:deviceId', function (request, respon
 // frontend UI
 app.set('jsonp callback name', 'cid');
 app.get('/getauthcode', function (req, resp) {
+  
+  /*forbid caching to force reload of getauthcode */
+  resp.set('Cache-Control','no-store, must-revalidate');
```
Comment (Fleker) [inline smart-home-provider/cloud/smart-home-provider-cloud.js]: Please add a space between arguments.

File: smart-home-provider/cloud/auth-provider.js (hunk 1 above)
Comment (orcema) [reply smart-home-provider/cloud/auth-provider.js]: done

File: smart-home-provider/cloud/datastore.js (hunk 2 above)
Comment (orcema) [reply smart-home-provider/cloud/datastore.js]: done

File: smart-home-provider/cloud/auth-provider.js (hunk 1 above)
Comment (orcema) [reply smart-home-provider/cloud/auth-provider.js]: This is just a copy of line 223, so i guess this one should be updated too?

**Comment by orcema** [comment]
#e335d2f
Couldn't update the settings. Check your connection  …
This happens when you try to link your account with a smartphone but you have ~~no devices defined for your account~~ you are not logged with a WebClient(Webbrowser). With this line of code if no devices are defined for your account,  the return value is an empty object properties

**Comment by Fleker** [comment]
I've tested it and haven't encountered any issues. I'll merge it and see if developers encounter any issues. Thanks a lot :+1: !

**Pull request merged**